scrape instagram and facebook / reddit, yahoo answers, hestenettet for statements on pain

In [1]:
pip install beautifulsoup4

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


TEXT GENERATION MODEL: https://huggingface.co/philschmid/flan-t5-base-samsum?text=hej

For this project, I am gonna use the Reddit API Praw. I am following this guide on how to implement it and parts of the code will be taken from here. https://towardsdatascience.com/scraping-reddit-data-1c0af3040768

In [2]:
pip install psaw

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [3]:
pip install praw

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 189.4/189.4 KB 15.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 55.9/55.9 KB 7.1 MB/s eta 0:00:00


In [4]:

import praw
from psaw import PushshiftAPI
import bs4
import requests
from bs4 import BeautifulSoup
from html.parser import HTMLParser
from praw.models import MoreComments

In [5]:
# authenticating myself!

reddit = praw.Reddit(client_id='Qbcp7XRs2X_HWLnshuyAAw',client_secret='yw9xGdrzGOn9wzE1qhVTZrPiAVa-5Q',user_agent='webscraper')

In [6]:
# i want to scrape from as many as possible subreddits containing discussions and threads on pain, illnesses etc. From there I need to iterate over all of them scraping the comments and subcomments.

In [7]:
# THIS IS WORKING!!!!!!
#i wanna make sure that the comments and subcomments I get with the keywords are actually about physical and mental pain, so I am limiting my search to a few subreddits. 
 
# part of code taken from: https://praw.readthedocs.io/en/latest/tutorials/comments.html

# list of key words
words_pain = ['pain','illnes','symptom','sick','treatment','diagnosed','undiagnosed','diagnosis','syndrome','drugs','drug','medication','meds','painkillers','disability','disease']

# list to hold scraped posts 
pain_all=[]

def get_comments():      
  for submission in reddit.subreddit('ChronicPain+ChronicIllness+backpain+disability+migraine+chronicpainrelief+kneepain+Fibromyalgia+painhub+painmed').hot(limit=None):          # iterating through subreddits
    submission.comments.replace_more(limit=None)          # iterate through all the comments 
    comment_queue = submission.comments[:]               # Seed with top-level
    id=submission.id
    title=str(submission.title)
    for word in words_pain:                              # iterate through keywords    
     if word in title or word in submission.comments: 
      while comment_queue:
        comment = comment_queue.pop(0)
        #print(comment.body)
        #comment_queue.extend(comment.replies) 
        # get both titles and bodies 
        pain_all.append(title)
        pain_all.append(comment.body)
  return pain_all

get_comments()
print(len(pain_all))
print(pain_all)



It is strongly recommended to use Async PRAW: https://asyncpraw.readthedocs.io.
See https://praw.readthedocs.io/en/latest/getting_started/multiple_instances.html#discord-bots-and-asynchronous-environments for more info.

It is strongly recommended to use Async PRAW: https://asyncpraw.readthedocs.io.
See https://praw.readthedocs.io/en/latest/getting_started/multiple_instances.html#discord-bots-and-asynchronous-environments for more info.

It is strongly recommended to use Async PRAW: https://asyncpraw.readthedocs.io.
See https://praw.readthedocs.io/en/latest/getting_started/multiple_instances.html#discord-bots-and-asynchronous-environments for more info.

It is strongly recommended to use Async PRAW: https://asyncpraw.readthedocs.io.
See https://praw.readthedocs.io/en/latest/getting_started/multiple_instances.html#discord-bots-and-asynchronous-environments for more info.

It is strongly recommended to use Async PRAW: https://asyncpraw.readthedocs.io.
See https://praw.readthedocs.io/en/l

2764
['I’m sick of trying to navigate a Healthcare system that doesn’t give a shit about me. ⚛️', 'where im at right now 🥲 im just so done', 'I’m sick of trying to navigate a Healthcare system that doesn’t give a shit about me. ⚛️', 'If only I had a choice other than deal with it or literally perish', 'I’m sick of trying to navigate a Healthcare system that doesn’t give a shit about me. ⚛️', 'We need them more than they need us :(', 'I’m sick of trying to navigate a Healthcare system that doesn’t give a shit about me.', 'It\'s exhausting having to stay on top of all of these "medical professionals" to make sure they do everything they say they are going To. And to question all their decisions because of the lack of communication. In the hospital they don\'t tell you much of anything they just do things without your consent and get angry thar you\'re questioning their actions', 'I’m sick of trying to navigate a Healthcare system that doesn’t give a shit about me.', "Omg it's infuriating

In [8]:
print('scraped content: ' ,pain_all)
print('amount of scraped content: ',len(pain_all))

scraped content:  ['I’m sick of trying to navigate a Healthcare system that doesn’t give a shit about me. ⚛️', 'where im at right now 🥲 im just so done', 'I’m sick of trying to navigate a Healthcare system that doesn’t give a shit about me. ⚛️', 'If only I had a choice other than deal with it or literally perish', 'I’m sick of trying to navigate a Healthcare system that doesn’t give a shit about me. ⚛️', 'We need them more than they need us :(', 'I’m sick of trying to navigate a Healthcare system that doesn’t give a shit about me.', 'It\'s exhausting having to stay on top of all of these "medical professionals" to make sure they do everything they say they are going To. And to question all their decisions because of the lack of communication. In the hospital they don\'t tell you much of anything they just do things without your consent and get angry thar you\'re questioning their actions', 'I’m sick of trying to navigate a Healthcare system that doesn’t give a shit about me.', "Omg it'

# DATA WRANGLING
Now we have a list of 2764 posts scraped from 10 different subreddits. 

As I want to be able to use this dataset for a variety of things; for finetuning and as inputs to different AI models, I want it to be cleaned up and labelled, but I still want to keep the essence intact, that meaning all emojis and special characters aswell as it being case-sensitive. 

In [9]:
import nltk 
from nltk.tokenize import sent_tokenize
import string
nltk.download('punkt')
nltk.download('averaged_perceptron_tagger')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Unzipping taggers/averaged_perceptron_tagger.zip.


True

In [18]:
# remove duplicates from list 
pain = [x for k,x in enumerate(pain_all) if x not in pain_all[:k]]

print('content with removed duplicates: ' ,pain)
print('number of comments: ',len(pain))

content with removed duplicates:  ['I’m sick of trying to navigate a Healthcare system that doesn’t give a shit about me. ⚛️', 'where im at right now 🥲 im just so done', 'If only I had a choice other than deal with it or literally perish', 'We need them more than they need us :(', 'I’m sick of trying to navigate a Healthcare system that doesn’t give a shit about me.', 'It\'s exhausting having to stay on top of all of these "medical professionals" to make sure they do everything they say they are going To. And to question all their decisions because of the lack of communication. In the hospital they don\'t tell you much of anything they just do things without your consent and get angry thar you\'re questioning their actions', "Omg it's infuriating to deal with any kind of Healthcare/Insurance anything! I would vote for ANYBODY that would give us Healthcare for all!!!!!", 'Backpain Thoracic Spine', "Where are you located? It doesn't sound like the US, so things might work a little differ

In [21]:
# hold cleaned dataset 
pain_newlines = []

# remove new lines 
for item in pain:
  pain_newlines.append(item.replace('\n',''))    

# hold cleaned dataset 
pain_clean = []

for item in pain_newlines:
  if 'http' in item: 
    continue
  else: 
    pain_clean.append(item)  

In [24]:
# label datasets 

pain_all=[]

# iterate over data to get text elements and index elements 
for index,text in enumerate(pain_clean):
  dictionary_all = {'text':text,'index':index}
  pain_all.append(dictionary_all)

# words that indicate a comment seeking advice
advice_seeking_words = ['do','how','?','which','what','who','when','whom','whose']

# hold the labeled data
data_labeled = []

# iterate over data 
for item in pain_all:
  text = item['text']
  tokens = nltk.word_tokenize(text)   
  pos_tags = nltk.pos_tag(tokens)

  if any([word in ['do','how','?','which','what','who','when','whom','whose'] for word,pos in pos_tags]):
   label = 1     # asking for advie
  else: 
    label = 0    # giving advice
  data_labeled.append({'text':text,'label':label})

len(data_labeled)


1511

In [27]:
# split into train and test 
pain_train = []
pain_test = []

pain_train.append(data_labeled[:1200])
pain_test.append(data_labeled[1200:])

# content is in 2d array - needs to be 1d 
pain_test = pain_test[0]
pain_train = pain_train[0]

print('test_len: ',len(pain_test))
print('train_len: ',len(pain_train))

test_len:  311
train_len:  1200


In [28]:
import json

# save data files
with open('data_pain_train.json','w',encoding='utf-8') as fout_train:   
  #fout.write(str(pain))
  json.dump(pain_train,fout_train)

with open('data_pain_test.json','w',encoding='utf-8') as fout_test: 
  json.dump(pain_test,fout_test)